llvm/lib/Transforms/Yk/BasicBlockTracer.cpp

```diff
@@ -0,0 +1,72 @@
+#include "llvm/Transforms/Yk/BasicBlockTracer.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+
+#define DEBUG_TYPE "yk-basicblock-tracer-pass"
+
+using namespace llvm;
+
+namespace llvm {
+void initializeYkBasicBlockTracerPass(PassRegistry &);
+} // namespace llvm
+
+namespace {
+struct YkBasicBlockTracer : public ModulePass {
+  static char ID;
+  Function *externalFunc = NULL;
+
+  YkBasicBlockTracer() : ModulePass(ID) {
+    initializeYkBasicBlockTracerPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool doInitialization(Module &M) override {
+    LLVMContext &Context = M.getContext();
+    if (externalFunc == NULL) {
```

Comment:
Should `externalFunc` be a field?

In theory you could run this pass twice on different modules and maybe end up mixing LLVM contexts.

Maybe `externalFunc` to `TraceFunc`, or `BlockRecFunc` or similar?

Reply:
You could also make it a local variable inside `runOnModule`.

Reply:
updated 👉 7195367457b04f536eb5fbf1d2c8fb7d621b676c